### PR TITLE
Address steady memory increase and bloated checkpoints

### PR DESCRIPTION
**Pull request by ksivaman**
Signed-off-by: Kirthi Shankar Sivamani <ksivamani@nvidia.com>

### CODE REVIEW

File: transformer_engine/pytorch/fp8.py [hunk 1]
```diff
@@ -65,22 +65,6 @@ def set_global_fp8_buffer(buffer: Dict[str, List[torch.Tensor]]) -> None:
...
-    """Sets global fp8 recompute buffer."""
-    global _fp8_tensors_recompute_buffer
-
-    # Map all tensors back to GPU.
-    for index, deck in enumerate(buffer):
-        buffer[index] = deque([[t.cuda() for t in tensors] for tensors in deck])
-
-    _fp8_tensors_recompute_buffer = buffer
-
-
```
Comment (ksivaman) [inline transformer_engine/pytorch/fp8.py:83]: Functions no longer needed

File: transformer_engine/pytorch/module.py [hunk 2]
```diff
@@ -209,7 +207,6 @@ def get_extra_state(self) -> torch.Tensor:
             state["scale_bwd"] = self.fp8_meta["scaling_bwd"].scale
             state["amax_history_bwd"] = self.fp8_meta["scaling_bwd"].amax_history
             state["global_fp8_buffer"] = get_global_fp8_buffer()
-            state["global_fp8_recompute_buffer"] = get_global_fp8_recompute_buffer()
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:212]: Omit storing this buffer in the checkpoint in order to reduce size

File: transformer_engine/pytorch/module.py [hunk 3]
```diff
@@ -269,11 +266,11 @@ def set_extra_state(self, state: torch.Tensor) -> None:
 
         # Restore global FP8 buffer states.
         set_global_fp8_buffer(state["global_fp8_buffer"])
-        set_global_fp8_recompute_buffer(state["global_fp8_recompute_buffer"])
-
         # Load extra items.
         self.fp8_meta.update(state["extra_fp8_variables"])
         self.fp8_meta["recipe"].amax_history_len = state["amax_history_fwd"].shape[0]
+        if "global_fp8_buffer_pos_fwd_recompute" in self.fp8_meta:
+            del self.fp8_meta["global_fp8_buffer_pos_fwd_recompute"]
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:273]: Now that we aren't storing the `_fp8_recompute_buffer` in the checkpoint, it is necessary that this key is not restored. If it is present, the tensor copy and restore functions in `fp8.py` for activation checkpointing will assume a non-empty buffer and index error will be thrown.

File: transformer_engine/pytorch/module.py [hunk 4]
```diff
@@ -452,6 +449,7 @@ def prepare_forward(
             # Activation recomputation is used and this is the first forward phase.
             if (
                 self.fp8
+                and self.training
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:452]: This check will omit copying the tensors for eval/inference case, which was the cause of increasing checkpoint size in the first place.

**Review by ptrendx — APPROVED** [review]
LGTM

**Comment by ptrendx** [comment]
I checked the CI failures and they are unrelated.

**Pull request merged**